Documentation/diagnostics/source_link.md

```diff
@@ -0,0 +1,137 @@
+        "C:\\src\\CodeFormatter\\src\\CodeFormatter\\CommandLinerParser.cs":    "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/CodeFormatter/CommandLineParser.cs",
+        "C:\\src\\CodeFormatter\\src\\DeadRegions\\Program.cs":                 "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/DeadRegions/Program.cs",
+        "C:\\src\\CodeFormatter\\src\\DeadRegions\\Options.cs":                 "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/DeadRegions/Options.cs",
+        "C:\\src\\CodeFormatter\\src\\DeadRegions\\OptionParser.cs":            "https://raw.githubusercontent.com/dotnet/codeformatter/bcc51178e1a82fb2edaf47285f6e577989a7333f/src/DeadRegions/OptionParser.cs",
+        ...
+    }
+}
+```
+
+Source Link JSON may contain multiple relative and/or absolute mappings in any order. They will be resolved in order from most specific to least specific. Here is an example of a valid Source Link JSON for an arbitrary project structure:
+```json
+{
+    "documents": {
+        "C:\\src\\*":                   "http://MyDefaultDomain.com/src/*",
+        "C:\\src\\foo\\*":              "http://MyFooDomain.com/src/*",
+        "C:\\src\\foo\\specific.txt":   "http://MySpecificFoodDomain.com/src/specific.txt",
+        "C:\\src\\bar\\*":              "http://MyBarDomain.com/src/*"
+    }
+}
+```
+In this example:
+- All files under directory `bar` will map to a relative URL beginning with `http://MyBarDomain.com/src/`.
+- All files under directory `foo` will map to a relative URL beginning with `http://MyFooDomain.com/src/` **EXCEPT** `foo/specific.txt` which will map to `http://MySpecificFoodDomain.com/src/specific.txt`.
+- All other files anywhere under the `src` directory will map to a relative url beginning with `http://MyDefaultDomain.com/src/`.
+
+## Tooling
+- [Visual Studio 2017](https://www.visualstudio.com/downloads/)
+    - First debugger to support Source Link.
+    - Only supports the [Portable PDB](https://github.com/dotnet/core/blob/master/Documentation/diagnostics/portable_pdb.md) format.
+    - The current implementation uses a case insensitive string comparison between the file path and the Source Link entry. Using multiple entries that differ only in casing is not supported; the first entry will be used and the second entry that differs only by case will be ignored. Here is an example of an unsupported Source Link JSON:
+        ```json
+        {
+            "documents": {
+                "/FOO/*": "http://.../X/*",
+                "/foo/*": "http://.../Y/*"
+            }
+        }
+        ```
+    - The current implementation does not support a Source Link entry that contains a URL that will cause a redirect. If the URL in the Source Link map causes a redirection, downloading the source file will fail.
```

Comment:
Is there a plan to fix that eventually? It looks like something that could potentially lead to big problems down the road since it's common practice to refactor web sites with permanent redirects to maintain old URL schemes. If GitHub were to do something like that, it would break debugging on lots of libraries.